### ISSUE TITLE
fltk 1.4.x current: hitting esc to close fltk app blows out the window manager

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#821]
**Describe the bug**
Just reporting what I'm seeing; I did a 'git pull' earlier today, rebuilt fltk as usual.

Now getting strange behavior I've not seen before with my apps AND the fltk test apps
(even the simple test/hello program) where, more often then not **hitting ESC to close an
FLTK app briefly freezes everything on the desktop, then the window manager restarts**,
bringing me back to the login screen.

Even **`test/hello`** exhibits this behavior.

Ubuntu 22 sees it as an "internal error" in **`/usr/bin/gnome-shell`** and does a whole crash report thing.

I still have an old build of fltk (from a day or two ago) with wayland disabled which doesn't seem to have this problem. Not sure if it's the recent git pull, or maybe ubuntu doing an upgrade, but so far it seems to be limited to recent FLTK built apps.

With my own app, I can open it and hit ESC and it closes OK, but if I reopen it and do a resize, then hitting ESC crashes the desktop.

With the 'hello' test program, hitting ESC almost always kills the desktop.
Strangely, hitting the 'x' so far has not caused the problem, and seems to quit normally.

**Comment by erco77** [comment]
Hmm, apparently I've done this enough times (about 20) where now gnome-terminal won't even open anymore.. yikes.

Had to install gvim just so I could run 'sudo apt install xterm' to get a terminal again.

So I did a test where I added a marker message to the /var/log/syslog with logger(1), then ran the test/hello program, hit ESC, and then after logging back in from the desktop crash, went into the syslog to see what I could find. Don't know if this is useful information, but apparently it got really pissed off, lol:

```
Nov  3 18:59:29 sabine erco: Hello world test: START
Nov  3 18:59:53 sabine gnome-shell[18514]: GNOME Shell crashed with signal 11
Nov  3 18:59:53 sabine gnome-shell[18514]: == Stack trace for context 0x55bd7dc63490 ==
Nov  3 18:59:54 sabine gnome-shell[20142]: (EE) failed to read Wayland events: Connection reset by peer
Nov  3 18:59:54 sabine update-notifier[20551]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine gnome-calendar[21052]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine nautilus[21952]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine snapd-desktop-i[18995]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine xdg-desktop-por[18926]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine xdg-desktop-por[19047]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine gsd-keyboard[18685]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine gsd-power[18688]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine gsd-media-keys[18687]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine gsd-wacom[18704]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine evolution-alarm[18753]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine gsd-color[18677]: Error reading events from display: Broken pipe
Nov  3 18:59:54 sabine systemd[1430]: org.gnome.SettingsDaemon.MediaKeys.service: Main process exited, code=exited, status=1/FAILURE
Nov  3 18:59:54 sabine kernel: [ 2811.441457] rfkill: input handler enabled
Nov  3 18:59:54 sabine MyFirefox.desktop[20902]: Exiting due to channel error.
Nov  3 18:59:54 sabine org.gnome.Nautilus[22199]: xterm: fatal IO error 11 (Resource temporarily unavailable) or KillClient on X server ":0"
Nov  3 18:59:54 sabine pulseaudio[18565]: X11 I/O error handler called
Nov  3 18:59:54 sabine MyFirefox.desktop[20955]: Exiting due to channel error.
Nov  3 18:59:54 sabine pulseaudio[18565]: X11 I/O error exit handler called, preparing to tear down X11 modules
Nov  3 18:59:54 sabine MyFirefox.desktop[20865]: Exiting due to channel error.
Nov  3 18:59:54 sabine systemd[1430]: org.gnome.Shell@wayland.service: Main process exited, code=dumped, status=11/SEGV
```
FWIW, "sabine" is the name of the machine.

**Comment by erco77** [comment]
Made this a high priority because it kills the entire desktop and all user applications, which is super bad.

**Comment by ManoloFLTK** [comment]
Tried here with Ubuntu 22.04 in VM on macOS:
- update fltk to last commit and build
- update Ubuntu to last package versions (linux-image version 5.15.0-88)
- reboot
- try many FLTK apps many times and close them with ESC

No trouble at all.

There were changes to the procedure behind window closing but that was weeks ago and it dealt with closing window with the mouse.

I've seen several times during development of the Wayland FLTK platform errors in the Wayland protocol that killed the desktop manager and returned to the login screen. I found that running an app with
`bin/test/hello >& hello.log &`
provided helpful information in the log file that can be fetched after having logged in again.

But, I see no reason why you would get protocol errors that I don't see here.

**Comment by erco77** [comment]
I'm about to end my programming day, but I'll try your redirecting stdout/err to a file and follow up with the results here.

**Comment by erco77** [comment]
Hmm, ok, after hours of easy replication earlier today, I can no longer replicate at the moment, so I'll have to revisit tomorrow.

My suspicion is, in trying to fix gnome-terminal after it wouldn't reopen after one of the many desktop crashes, renaming away my home directory files may have affected the problem.

I do have a couple of custom window manager behaviors I insist on having;

- focus follows mouse
- disabling edge tiling (touching a window at top of screen maximizes -- drives me crazy when doing UI layouts!)

..and maybe one or two others, can't recall.

So it's possible the problem is purely in the window manager with custom settings applied.

I still have my home dir files from earlier today, so I'll try to re-investigate tomorrow; too tired now.

**Comment by erco77** [comment]
OK, I didn't think I could get to sleep unless I tried restoring my home directory, so I did by logging out, sshing in from a different machine, renaming away my "working" home files, put back my "bad" home files, then walked back to the machine, logged into the desktop, and now can replicate the crashing problem again.

I did try the redirect stdout/err thing to a file, but after the crash the file was just zero bytes in size, so no error messages were printed by FLTK at the time of the crash.

I'm going to kick off, but if you can think of anything you'd like me to try tomorrow in the way of debugging, I can try to supply it. I might be able to attach gdb to gnome-shell process from an ssh terminal as root to see what kills it via a stack trace.

I'm guessing FLTK isn't printing anything, and just whatever it does during an ESC that's different from 'x' is tripping a bug in gnome-shell, perhaps when certain window manager flags are set.

I didn't set many, but the ones I named I did for sure, using either dconf-editor or gnome-tweaks.

**Comment by Albrecht-S** [comment]
> I didn't set many flags, but the ones I named I did change for sure, using either dconf-editor or gnome-tweaks or both.

@erco77 If you could tell us the features you changed together with detailed instructions on how to change them someone could maybe try to do the same in a VM to reproduce the effect. Can you also provide more details about your system environment? I see and/or guess you're using:
- Ubuntu 22 (maybe 22.04 as Manolo used in his VM, or which one?)
- Desktop Gnome/Wayland ?
- Is your system up-to-date? (apt update;apt upgrade;)
- Which kernel version? (uname -a)
- Any other info that might be useful?

**Comment by erco77** [comment]
It's an updated/upgraded Ubuntu 22.04.3 LTS system as of two days ago. (To install opengl they recommended an apt update+upgrade). Output of uname -a is:

    Linux sabine 6.2.0-36-generic #37~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC
    Mon Oct  9 15:34:04 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Desktop is the default I think, the process table shows gnome-shell/gdm-wayland-session and Xwayland.
The Settings -> About says Windowing System is Wayland, GNOME version 42.9, 64bit.

It's running native on a mac mini (no VM)
Graphics is Mesa Intel UHD Graphics 630 (CFL GT2).

I'm not sure I can exactly describe the settings; I'd just google how to do each as I encountered things I didn't like.
There weren't many, but I don't think they could be done with the default window manager "Settings"; it provides very little

- To disable edge tiling: dconf-editor was used: /org/gnome/mutter/edge tiling (OFF)
- To enable focus follows mouse: gnome-tweaks was used: Windows > Window Focus > Focus on Hover
- To disable "Attach Modal Dialogs": gnome-tweaks was used: Windows > Attach Modal Dialogs (OFF)
- To enable day of week in top bar: gnome-tweaks: Top Bar -> Clock -> Weekday (ON)

I don't think there was much more than that. But I don't know if those settings alone cause the problem, I can try it with a different account (switch user).

I suppose I could tar off my home directory, so that one of you can create the 'erco' user, and extract my tar into /home/erco, then chown/chmod it (since when you add the user, it'll probably have a different uid/gid). That's kinda how I'm replicating things.

**Comment by erco77** [comment]
It looks like `dconf dump /` is what I needed; dconf apparently has its own file system lol. So now I can compare the ASCII of the erco user's dconf stuff to the new user in the switch account to see if I can replicate with the new user. It's only about 200 lines of text, so might be able to zero in on if that's the cause.

**Comment by erco77** [comment]
OK, wow, yeah, that worked; I can replicate the crash in the new user account, which means it's been isolated.
What I did:

1. As the user erco: 'dconf dump / > /var/tmp/erco.txt'
2. Changed the one instance of 'erco' within that file to the new user account's name
3. Switched user to the new user's account's desktop
4. As the new user: 'dconf load / < /var/tmp/erco.txt'
5. Ran the fltk hello program, moved the window a little bit to a new position, hit ESC, and that replicated the crash.

Attaching my erco.txt file here to see if anyone else can replicate the crash.

I advise you create a different account as I did, just be sure to edit the file first and replace the string 'erco' with the new user name, before doing the dconf load. I used the `adduser` command to create the new account.
[erco.txt](https://github.com/fltk/fltk/files/13256779/erco.txt)

**Comment by erco77** [comment]
Another thing I noticed, perhaps related:

1. Run a couple of FLTK apps (hello + checkers) and leave them running, along with some non-fltk apps
2. Click the top right of the desktop (power button icon) Logout > Switch User > and just log back in as the same user
3. The FLTK apps closed (!) showing this error in the terminal that invoked them:
```
      Fatal error while communicating with the Wayland server: Invalid argument
```
This seems to happen even with a default user login, none of the dconf stuff mentioned in the comments above.
The non-fltk apps remain open, unaffected.

**Comment by erco77** [comment]
So to summarize, devs should be able to replicate the crash by doing:

1. Save [dconf.txt](https://github.com/fltk/fltk/files/13260079/dconf.txt) on your machine in /var/tmp/dconf.txt
2. Create a new user account using `sudo adduser someguy`
3. Login as that new user
4. Open a terminal and run: `dconf load / < /var/tmp/dconf.txt`, which applies the desktop settings I've been using.
5. Run fltk's `test/hello`
6. Move the window a little bit by dragging its titlebar (this helps cause the crash, dunno why)
7. Hit ESC, and in my case it hangs then resets the window manager, apparently tripping a coredump in `gnome-shell`

This could be a bug in gnome-shell we're tripping, but might seems odd exiting with 'x' and/or other apps not doing this.

Also, as noted in the comment above this one, there's a possibly related problem where FLTK apps close when you return to the desktop from 'switch user'.

**Comment by ManoloFLTK** [comment]
@erco77 Many thanks for the detailed analysis described above. I just tried here under Debian 12 (GNOME Shell 44.5), and the infamous `focus-change-on-pointer-rest=false` doesn't trigger crash. I will try now with Ubuntu 22.04.

**Comment by ManoloFLTK** [comment]
I just tried with Ubuntu 22.04 inside a VM on an M1 MacBook Pro (GNOME Shell 42.9) and get no crash after doing
```
cat > crash.txt
[org/gnome/mutter]
focus-change-on-pointer-rest=false
ctrl-D

dconf load / < crash.txt

dconf dump / | grep focus
focus-change-on-pointer-rest=false

test/hello
    move window, hit ESC
```

Is the issue architecture-dependent (x86_64 vs arm64)?
Are VMs immune from the issue?

**Comment by ManoloFLTK** [comment]
To reproduce the crash with Ubuntu 22 (GNOME Shell 42.9), I need here
```
dconf dump / | grep focus
focus-mode='sloppy'
focus-change-on-pointer-rest=false
```
If I set focus-mode back to `click`with `gnome-tweaks`, the crash disappears.
If I remove the `focus-change-on-pointer-rest` key, the crash disappears.

No crash under same dconf settings with Debian 12 (GNOME Shell 44.5).

No crash under same dconf settings with Ubuntu 23.10 (GNOME Shell 45.0).

My view on this issue: 
- this is a bug in gnome 42 corrected in gnome 44, not really a bug in FLTK. 
- @erco77 reports above how to work around the issue while still using `focus-mode='sloppy'` by removing key `focus-change-on-pointer-rest`.

I will investigate a bit more to search if the crash in gnome 42 can be avoided by working in FLTK.

**Comment by ManoloFLTK** [comment]
Under the `dconf` settings that promote the crash of gnome 42, using `tcsh` as shell, and doing this
`(bin/test/hello ; echo $status) >& hello.log&`
I find that `test/hello` goes all the way to return 0 from `main()` after having processed the esc key normally for an FLTK app, and returns 0 as status code to the shell.

Therefore, the crash occurs after a normal end from the FLTK app. I don't see what to do to prevent this crash.

**Comment by erco77** [comment]
I see, OK I guess Ubuntu 22 must just be buggy then; I seem to keep hitting things specific to it.

We can close if you think there's nothing we can do.
I don't suppose changing what happens when hitting ESC could help? Like calling _exit(0) or some such.
Or maybe there's some order as to how the wayland connection is shutdown, or perhaps some message we should be emitting to the window manager (the desktop protocols seem to keep changing all the time).

It's just odd other apps don't seem to trigger this, and from what I can tell, neither does the 'x' title bar button on fltk apps. Not really sure how to report it as a bug against the window manager other than to letting the Ubuntu crash reporter speak for itself.. I've been letting those go through now.

Also, this is an LTS (Long Term Support) version of Ubuntu which a lot of my customers might be using, so I'm a little worried how much weirdness they'll run into of they're on this release.

I should probably just upgrade to Ubuntu 23..

**Comment by ManoloFLTK** [comment]
I will try (tomorrow at best) to close the Wayland connexion upon return from main(). May be GTK apps do that. FLTK doesn't for now.

I tried a modified hello app with 2 identical windows. The crash occurs when the 2nd displayed window is closed (by ESC) first and the other second. No crash when the windows are closed (by ESC) in the other order. This really looks like a Mutter bug. It also shows that the first window closing operation leaves the Wayland communication in good shape because the remaining window runs OK.

**Comment by ManoloFLTK** [comment]
Closing the Wayland connexion with `fl_close_display()` just before returning from `main()` doesn't protect from the crash. I see nothing more to be done from inside FLTK.

> Also, this is an LTS (Long Term Support) version of Ubuntu which a lot of my customers might be using, so I'm a little worried how much weirdness they'll run into of they're on this release.

The issue arises only when focus is set to follow mouse and there's a solution to avoid the crash also in this setting.

**Comment by ManoloFLTK** [comment]
I think the commit mentionned above fixes the issue.

This commit makes `fl_open_display()` add an `atexit` function to all FLTK client apps. This function calls `wl_display_roundtrip()` which forces all compositor-to-client messages pending in the socket to be read and processed by the client app. This makes the compositor fully informed that one of its client apps has terminated, which is apparently necessary to avoid the issue.

@erco77 Please, close issue if you agree it's fixed.

**Comment by Albrecht-S** [comment]
@ManoloFLTK Awesome, great findings and great fix. Thanks from me. Hopefully it does really fix @erco77's issue in his constellation.

**Comment by erco77** [comment]
Great - I'll give it a try and follow up with a close if it fixes.
I had to stop using that machine for a while as it was too easy to blow everything up by hitting ESC - something us VI guys do often :^D

**Comment by erco77** [comment]
OK, looks good!

I've been using the problem machine all day, and haven't had a single desktop blow up, so I think you fixed it! Closing..

**Comment by ManoloFLTK** [comment]
> Another thing I noticed, perhaps related -- fltk apps disappear off the desktop when returning from switch user:
> 
>     Run a couple of FLTK apps (hello + checkers) and leave them running, along with some non-fltk apps
>     Click the top right of the desktop (power button icon) Logout > Switch User > and just log back in as the same user
>     The FLTK apps closed (!) showing this error in the terminal that invoked them:
> 
>       Fatal error while communicating with the Wayland server: Invalid argument

@erco77 This is now fixed with 93b26e4 (mostly) and 5989caf (a bit more).

**Comment by erco77** [comment]
Great -- I haven't checked but will tomorrow.. I'm doing hardware today.
